[CORE-3183] schema registry json compatibility checks for primitive types

andijcr opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Implementation of `bool check_compatible(const json_schema_definition& reader, const json_schema_definition& writer);` 
for string, numeric JSON types, to allow checking compatibility of a subset of JSON schema draft 4
  
The checking is done by traversing both schemas, looking for a rule in the writer schema that is less strict than its counterpart in the reader schema. 
If no such rule is found, then the result is true

Fixes: https://redpandadata.atlassian.net/browse/CORE-3183

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.1.x
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

BenPope reviewed (commented):
Looks ok in principal.

1. Maybe a `std::bitset` is a nicer than a `std::set<enum>`

pgellert commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -298,6 +300,454 @@ result<json::Document> parse_json(std::string_view v) {
 
+/// is_superset section
+
+// a schema O is a superset of another schema N if every schema that is valid
+// for N is also valid for O. precondition: older and newer are both valid
+// schemas
+bool is_superset(json::Value const& older, json::Value const& newer);
+
```
I think this forward declaration is not necessary

andijcr replied:
yes it is right now. I put it here to implement the types that require recursion, like object and array. initially i thought to implement is_array_superset here but i decided to leave it for another pr

pgellert commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -298,6 +300,454 @@ result<json::Document> parse_json(std::string_view v) {
+        std::ranges::copy(
+          type_it->value.GetArray()
+            | std::views::transform([](json::Value const& v) {
+                  return boost::lexical_cast<json_type>(
+                    std::string_view{v.GetString(), v.GetStringLength()});
+              }),
+          std::back_inserter(ret));
+    } else {
+        ret.push_back(boost::lexical_cast<json_type>(std::string_view{
+          type_it->value.GetString(), type_it->value.GetStringLength()}));
```
nit: the parsing of a `json::Value` into a `json_type` is duplicated here

andijcr replied:
what do you mean?

BenPope replied:
The cast, I think, so something like my suggested `parse_json_type` would result in:
```c++
        std::ranges::copy(
          type_it->value.GetArray() | std::views::transform(&parse_json_type),
          std::back_inserter(ret));
    } else {
        ret.push_back(parse_json_type(type_it->value));
    }
```

andijcr replied:
done

pgellert commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -298,6 +300,454 @@ result<json::Document> parse_json(std::string_view v) {
+        }
+    }
+
+    if (auto [maybe_gate_value, older_it, newer_it]
```
nit: there's some code duplication here across `exclusiveMinimum` and `exclusiveMaximum`. I think the only difference is the check on the boolean values and the error message. Reusing the rest of the code would make it clearer that that's the only difference.

pgellert commented on src/v/pandaproxy/schema_registry/test/test_json_schema.cc:
```diff
@@ -124,3 +124,179 @@ SEASTAR_THREAD_TEST_CASE(test_make_valid_json_schema) {
+      .writer_schema = R"({"type": "string", "pattern": "^test"})",
+    },
+  });
+SEASTAR_THREAD_TEST_CASE(test_compatible_schemas) {
```
nit: This looks like a copy of `test_incompatible_test_cases` with a boolean value negated. I'd probably merge the two tests to reduce duplication. You could even reuse the test case struct types.

andijcr replied:
caught red handed

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -423,11 +460,94 @@ bool is_superset(json::Value const& older, json::Value const& newer) {
+           "oneOf",
+           "not",
+         }) {
+        if (
+          newer.HasMember(not_yet_handled_keyword)
+          || older.HasMember(not_yet_handled_keyword)) {
+            // these keyword are not yet handled, their presence might change
+            // the result of this function
+            throw std::runtime_error{fmt::format(
+              "{} not fully implemented yet. unsupported keyword: {}, input: "
+              "older: '{}', newer: '{}'",
+              __FUNCTION__,
+              not_yet_handled_keyword,
+              pj{older},
+              pj{newer})};
+        }
```
> I'm assuming the plan here is to reduce not_yet_handled_keyword to an empty list in follow-up PRs, right?
yes correct
> Can the users add keywords that we don't know about?
No, the set of keywords in the JSON schema is closed, and in this function, the JSON has already passed the test of being a valid schema.
What can happen is that the user is reusing a keyword in a type that ignores it, for example, `{"type": "integer," "format": "rm -rf /"},` the format is a keyword that has meaning only for "type": "string." In this case, until we remove the keyword from this list, there is a risk of rejecting a valid schema.
Eventually, we should remove this list altogether in the span of some PR

BenPope reviewed (commented):
I didn't get all the way through this.

In general I'd prefer it if the error handling didn't throw `std::runtime_error`, and instead uses the `as_exception(error_info{})` pattern.

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -298,6 +299,86 @@ result<json::Document> parse_json(std::string_view v) {
+    null = 6
+};
+// enough inlined space to hold all the values of json_type
+using json_type_list = absl::InlinedVector<json_type, 7>;
```
I want to express the set difference operation (the first op performed in is_superset), and for that, I can use std::set_difference. This requires an ordered iterable, and the options would be 
- std::set   - allocation for just 7 elements but uniqueness order guarantee
- absl::InlinedVector - inlined storage, need to manually perform sorting (uniqueness is guaranteed by metaschema validation)
- boost::dynamic_bitset - dynamic storage but has bitset semantics and iterator support

and I also want to iterate over the values to run the specific is_[numerics|string|array|object]_superset

std::bitset lacks iterator support, I tried to use it but i don't think it's an improvement 
`auto newer_minus_older = newer_types & older_types.flip()`

what do you suggest?

BenPope replied:
I guess `InlinedVector` is not worse than most of the alternatives.

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -298,6 +299,86 @@ result<json::Document> parse_json(std::string_view v) {
+              }),
+          std::back_inserter(ret));
+    } else {
+        ret.push_back(boost::lexical_cast<json_type>(std::string_view{
```
Exceptions are not possible here, because the metaschema validation already checked that the list of types contains only allowed words

BenPope replied:
I was assuming `lexical_cast` used `std::istream& operator>>(std::istream& is, json_type& t)` which sets the `fail_bit`, which would result in a throw of `boost::lexical_cast_exception`, so the goal was to throw a suitable exception if it happened. 

If it's impossible, then the parse will always return an engaged optional and the semantics are preserved.

andijcr replied:
ok

andijcr commented:
force push: addressed most of the review comments
- changed std::runtime_error for as_exception(invalid_schema
- simplified unit test case
- from_string_view to_string_view for json_type
- exclusiveMinimum/exclusiveMaximum simplification

BenPope commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -411,6 +411,89 @@ json_type_list normalized_type(json::Value const& v) {
+    // 0.1 is a valid json literal, but does not have an exact double
+    // representation. this cannot be caught with this, and it would require
+    // some sort of decimal type
+    if (!older_val_p->IsLosslessDouble() || !newer_val_p->IsLosslessDouble()) {
```
It might be worth coming back to this.

andijcr replied:
definetely

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -395,6 +395,89 @@ json_type_list normalized_type(json::Value const& v) {
+        return std::nullopt;
+    }();
+
+    return {gate_check, &older_it->value, &newer_it->value};
```
you are right, silly mistake. originally i returned the iterator but then i refactored this blindly

andijcr commented:
force push: other review comments

BenPope commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -298,6 +298,97 @@ result<json::Document> parse_json(std::string_view v) {
+        }
+    } else {
+        ret.push_back(parse_json_type(
+          {type_it->value.GetString(), type_it->value.GetStringLength()}));
```
Ooops.

pgellert reviewed (dismissed):
Apart from Ben's comment of using `parse_json_type(type_it->value)` inside `normalized_type`, it lgtm.

andijcr commented:
fixed unnecessary json::Value copy in normalized_type
